Makefile

```diff
@@ -43,22 +43,36 @@ build/man/%: doc/%.pod .version | build/man
 		$< $@
 
 test: check # just a synonym
-check: default test/cov.html test/cov.func.txt
-test/cov.cov: clean-tests build/holo.test
+.PHONY: test
+
+check: default
+check: check-gofmt
+check: check-golint
+check: test/cov.html test/cov.func.txt
+.PHONY: check
+
+check-gofmt:
 	@if s="$$(gofmt -l $(go_srcs) 2>/dev/null)" && test -n "$$s"; then printf ' => %s\n%s\n' gofmt  "$$s"; false; fi
+check-golint:
 	@if s="$$(golint   $(go_dirs) 2>/dev/null)" && test -n "$$s"; then printf ' => %s\n%s\n' golint "$$s"; false; fi
-	@$(GO) test $(GO_TESTFLAGS) -coverprofile=test/cov/holo-output.cov $(pkg)/cmd/holo/internal
-	@$(GO) test $(GO_TESTFLAGS) -coverprofile=test/cov/ssh-keys-output.cov $(pkg)/cmd/holo-ssh-keys/impl
+%/go-test.cov: clean-tests
+	@$(GO) test $(GO_TESTFLAGS) -coverprofile=$@ $(pkg)/$*
+check-holo-test-help: clean-tests build/holo.test util/holo-test-help
 	@HOLO_BINARY="$$PWD/build/holo.test" HOLO_TEST_COVERDIR=$$PWD/test/cov ./util/holo-test-help
+test/%/check: test/%/source-tree clean-tests build/holo.test util/holo-test
 	@\
 		export HOLO_BINARY=../../../build/holo.test && \
 		export HOLO_TEST_COVERDIR=$(abspath test/cov) && \
 		export HOLO_TEST_SCRIPTPATH=../../../util && \
-		$(foreach p,files run-scripts ssh-keys users-groups,\
-			ln -sfT ../build/holo.test test/holo-$p && \
-			./util/holo-test holo-$p $(sort $(wildcard test/$p/??-*)) && ) \
-		true
-	util/gocovcat.go test/cov/*.cov > test/cov.cov
+		ln -sfT ../build/holo.test test/holo-$(firstword $(subst /, ,$*)) && \
+		./util/holo-test holo-$(firstword $(subst /, ,$*)) $(@D)
+.PHONY: check-% test/%/check
+
+test/cov.cov: $(sort $(shell find $(filter-out %.go,$(go_srcs)) -name '*_test.go' -printf '%h/go-test.cov\n'))
```

Comment:
This may be of interest to you: https://github.com/sapcc/limes/blob/ab4245a8f195672b808f990f9f7c2452652c475d/Makefile#L22